Feature/arsn 285 modify getObjectMdStats() to collect more data metrics

XinLi0207 opened the pull request:
the design is here: https://github.com/scality/citadel/blob/development/1.0/docs/design/storage-consumption-metrics.md

this PR mainly modified `getObjectMDStats()` output from 
```
        "objects" : 4,
        "versions" : 6,
        "buckets" : 3,
        "bucketList" : [ 
            {
                "name" : "bucket2",
                "location" : "us-east-1",
                "isVersioned" : true,
                "ownerCanonicalId" : "2fbfcc0125094acda54f0cabb5d1e66ae78aaf67f6060e8ac5b6abc9ada01d81",
                "ingestion" : false
            }, 
            {
                "name" : "bucket1",
                "location" : "us-east-1",
                "isVersioned" : true,
                "ownerCanonicalId" : "2fbfcc0125094acda54f0cabb5d1e66ae78aaf67f6060e8ac5b6abc9ada01d81",
                "ingestion" : false
            }
        ],
        "dataManaged" : {
            "total" : {
                "curr" : 392,
                "prev" : 222
            },
            "byLocation" : {
                "us-east-1" : {
                    "curr" : 392,
                    "prev" : 222
                }
            }
        },
        "stalled" : 0,
```
to
```
{
        "objects" : 4,
        "versions" : 6,
        "buckets" : 3,
        "bucketList" : [ 
            {
                "name" : "bucket2",
                "location" : "us-east-1",
                "isVersioned" : true,
                "ownerCanonicalId" : "2fbfcc0125094acda54f0cabb5d1e66ae78aaf67f6060e8ac5b6abc9ada01d81",
                "ingestion" : false
            }, 
            {
                "name" : "bucket1",
                "location" : "us-east-1",
                "isVersioned" : true,
                "ownerCanonicalId" : "2fbfcc0125094acda54f0cabb5d1e66ae78aaf67f6060e8ac5b6abc9ada01d81",
                "ingestion" : false
            }
        ],
        "dataManaged" : {
            "total" : {
                "curr" : 392,
                "prev" : 222
            },
            "byLocation" : {
                "us-east-1" : {
                    "curr" : 392,
                    "prev" : 222
                }
            }
        },
        "stalled" : 0,
        "dataMetrics" : {
            "bucket" : {
                "bucket1" : {
                    "usedCapacity" : {
                        "current" : 376,
                        "nonCurrent" : 0
                    },
                    "objectCount" : {
                        "current" : 2,
                        "nonCurrent" : 0,
                        "deleteMarker" : 0
                    }
                },
                "bucket2" : {
                    "usedCapacity" : {
                        "current" : 6,
                        "nonCurrent" : 0
                    },
                    "objectCount" : {
                        "current" : 1,
                        "nonCurrent" : 0,
                        "deleteMarker" : 0
                    }
                },
            },
            "location" : {
                "us-east-1" : {
                    "usedCapacity" : {
                        "current" : 392,
                        "nonCurrent" : 222
                    },
                    "objectCount" : {
                        "current" : 4,
                        "nonCurrent" : 6,
                        "deleteMarker" : 1
                    }
                }
            },
            "account" : {
                "account2" : {
                    "usedCapacity" : {
                        "current" : 392,
                        "nonCurrent" : 222
                    },
                    "objectCount" : {
                        "current" : 4,
                        "nonCurrent" : 6,
                        "deleteMarker" : 1
                    }
                }
            }
        }
    }
}
```

added 3 metric levels: bucket/location/account and collect their `usedCapacity` and `objectCount` respectively

Commit:
feature: ARSN-285 modify getObjectMDStats() to include metrics needed for storage consumption

bert-e commented:
# Hello xinliscality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/arsenal/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

Commit:
feature: ARSN-285 getObjectMDStats functional tests

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ARSN-285 contains:


* `8.1.76`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.1.77`


Please check the `Fix Version/s` of ARSN-285, or the target
branch of this pull request.

Commit:
feature: ARSN-285 update scanItems()

williamlardier reviewed (commented):
First iteration, reviewed:
- code (most of my comments)
- consistency of computed metrics with previous computations, that looks good to me

Next review will focus more on usage of the changes along with s3utils job.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

alexanderchan-scality reviewed (commented):
these changes could be made in `s3utils` by extending the `mongoclientinterface` class. reason for doing so  would be to: 
1. move the counting logic closer to the count script 
2. separate the count logic from the s3 op mongoclient in arsenal

changes to `consolidateData` and `scanItemCount`  won't be needed; these  method have been deprecated and are not   in the code path of  the count items script

Commit:
feature: ARSN-285 post comment 1

XinLi0207 commented:
Hi @alexanderchan-scality 
thanks for reviewing, yeah you are right, `scanItems()` and `consolidate()` here are not used anywhere, we have [those logics in s3utils](https://github.com/scality/s3utils/pull/276/files#diff-370b727cbc9f2bd393b509e0c010aea03b5957fdacf07e063c8a75c07f58b4cfR69) directly. To reduce complexity, I just removed these two functions, since they are not used anywhere.

> these changes could be made in s3utils by extending the mongoclientinterface class.

here **these changes** you meant only `consolidateDate` and `scanItemCount` or the whole `getObjectMDStats()` and its related util functions?

XinLi0207 commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -1898,24 +1904,51 @@ class MongoClientInterface {
-            const locations = dataManaged.locations;
-            store.dataManaged.total.curr += dataManaged.total.curr;
-            store.dataManaged.total.prev += dataManaged.total.prev;
+        if (!results) return;
```
removed the whole `consolidateDate()` function as it's deprecated and not used anywhere

the same logic is implemented in [s3utils](https://github.com/scality/s3utils/pull/276/files#diff-370b727cbc9f2bd393b509e0c010aea03b5957fdacf07e063c8a75c07f58b4cfR69) directly, to be more close with countItem

XinLi0207 commented on lib/storage/metadata/mongoclient/MongoClientInterface.js:
```diff
@@ -1898,24 +1904,51 @@ class MongoClientInterface {
+                    } else {
+                        store.dataMetrics[metricLevel][resourceName] = results.dataMetrics[metricLevel][resourceName];
+                    }
+                });
```
removed the whole consolidateDate() function as it's deprecated and not used anywhere

the same logic is implemented in [s3utils](https://github.com/scality/s3utils/pull/276/files#diff-370b727cbc9f2bd393b509e0c010aea03b5957fdacf07e063c8a75c07f58b4cfR69) directly, to be more close with countItem

will optimize it in consolidateData() function in s3utils

alexanderchan-scality commented:
@XinLiScality 

> here these changes you meant only consolidateDate and scanItemCount or the whole getObjectMDStats() and its related util functions?

sorry for not being clear. what I had meant was that the changes that were made for `getObjectMDStats()`; it could be a method overwrite in s3utils.

`consolidateData` and `scanItemCount` could be left untouched as there looks to be some tests in cloudserver that calls on it; this will have to be left as a tech debt ticket: to remove the deprecated method and the test in cloudserver

XinLi0207 commented:
hi @alexanderchan-scality,

I implemented a new `CountItemsMongoClient` in [s3utils](https://github.com/scality/s3utils/pull/276) which extends MongoClientInterface  and overrides getObjectMDStats() and some other functions, I agree with what you said, it's indeed better to be closer to CountItems logic and should not be a common mongo client function.

if the s3utils PR looks good to you, there is nothing we need to change in arsenal, I will close this PR then
I have also moved all the tests I have done in this https://github.com/scality/Arsenal/pull/2055 here as well, so now there is no change need to be done at arsneal side, I will close that PR if this PR looks good to you.

XinLi0207 commented:
close this PR as work is done in https://github.com/scality/s3utils/pull/276 as an extension of mongoClientInterface class
